### PR TITLE
testAbsoluteJointSpaceJump: improve readability of code

### PR DESCRIPTION
**Pull request by rhaschke**
I separated the source code cleanup of `testAbsoluteJointSpaceJump` from #850 to allow for early merging of the latter and thus fixing the ROS build farm.

@mlautman, @davetcoleman Please argue why do you think that this cleanup is not more readable.
I think, reducing the number of `joint->getType()` checks improves readability.

### CODE REVIEW

File: moveit_core/robot_state/src/robot_state.cpp [hunk 1]
```diff
@@ -2039,28 +2039,40 @@ double RobotState::testRelativeJointSpaceJump(const JointModelGroup* group, std:
 double RobotState::testAbsoluteJointSpaceJump(const JointModelGroup* group, std::vector<RobotStatePtr>& traj,
                                               double revolute_threshold, double prismatic_threshold)
 {
-  bool test_revolute = revolute_threshold > 0;
-  bool test_prismatic = revolute_threshold > 0;
+  struct
+  {
+    double limit;
+    bool check;
+  } data[2] = { { revolute_threshold, revolute_threshold > 0.0 }, { prismatic_threshold, prismatic_threshold > 0.0 } };
```
Comment (davetcoleman) [inline moveit_core/robot_state/src/robot_state.cpp]: this struct is overly-complicated and takes much longer to understand than two simple bools in the original approach. 

- you bundle declaring the struct, creating an array of structs, and initializing them unnecessarily, which takes a long time to parse
- you mix in conditional logic with the initializer 
- there is an unnecessary copy of threshold values

**Comment by rhaschke** [comment]
@davetcoleman You seem to complain only about the struct, don't you?

> This struct is overly-complicated and takes much longer to understand than two simple bools in the original approach.

I think the struct is simple, but you are right, I'm combining declaring, instantiating, and initializing the struct.
I think this is usual practice in C++ programming and the most efficient way to initialize.
Do you prefer to split these three steps?

> you bundle declaring the struct, creating an array of structs, and initializing them unnecessarily, which takes a long time to parse

I guess you refer to human "parsing". The compiler probably doesn't care.

> you mix in conditional logic with the initializer

Yes, why not? First computing the bool and then assigning to the struct would require copying - which you complained about for the other variables.

> there is an unnecessary copy of threshold values

I create this struct (and need to copy the threshold values, yes) because I want to index into the struct array and access both independent variables in a _uniform_ fashion. For me, it's much easier to "parse" the struct array definition than the nested and repeated `if` conditions of the previous code. But maybe that's a matter of taste.

**Comment by davetcoleman** [comment]
Can you declare the struct then initialize it in two separate statements / lines? I think this would be enough to be more human readable.

**Comment by rhaschke** [comment]
@davetcoleman Done your requested changes and rebased due to conflicts.

**Comment by davetcoleman** [comment]
Travis reports build errors though:
```
std::vector<std::shared_ptr<moveit::core::RobotState> >&, double, double)::LimitData data [2]’ have incomplete type
                         { prismatic_threshold, prismatic_threshold > 0.0 } };
                                                                            ^
/root/ws_moveit/src/moveit/moveit_core/robot_state/src/robot_state.cpp:2048:76: error: storage size of ‘data’ isn’t known
```

**Comment by mlautman** [comment]
I strongly believe that these changes shouldn't be merged. This takes really simple code and adds data types, a switch statement, and control flow that simply does not need to be there. Code should be simple not complex.

**Comment by rhaschke** [comment]
I think that code should be easily readable as well as efficient. The set of lengthy boolean expressions in the previous code were much harder to comprehend than the simple switch case statements here. I agree that my code is longer, but in my opinion it is easier to read.

**Comment by v4hn** [comment]
In my humble opinion this whole request is bikeshedding.
If Robert insists on it (which he did), then it was ok to merge.
There was one approval already.

**Comment by v4hn** [comment]
> Please refrain from calling bikeshedding to dismiss opposing opinions when you get bored of them.

That was not my intention and I hope @mlautman did not understand it that way. :fearful:
I called bikeshedding on the whole request and personally I agree with @mlautman and your statements.

But given support from two maintainers it was ok to merge the request.
And discussing a request to change it back without functional changes or wider scope is not any better from my perspective..

Thanks for the sweets. :smile:

**Comment by mlautman** [comment]
@v4hn I think we're on the same page here :)

**Comment by davetcoleman** [comment]
To clarify my PR approval - it was a compromise where @rhaschke separated out the creation of the struct from its initialization, which was a step towards the readability @mlautman argues. Overall I wanted to move on with this new feature so approved it.

**Pull request merged**